cmd/cluster-controller/main.go

```diff
@@ -61,27 +72,50 @@ func main() {
 		syncerMode = cluster.SyncerModePush
 	}
 
-	clusterController, err := cluster.NewController(r, *syncerImage, kubeconfig, resourcesToSync, syncerMode)
+	kcpSharedInformerFactory := kcpexternalversions.NewSharedInformerFactoryWithOptions(kcpclient.NewForConfigOrDie(r), resyncPeriod)
+	crdSharedInformerFactory := crdexternalversions.NewSharedInformerFactoryWithOptions(apiextensionsclient.NewForConfigOrDie(r), resyncPeriod)
+
+	apiResourceClient := typedapiresource.NewForConfigOrDie(r)
+	crdClient := typedapiextensions.NewForConfigOrDie(r)
+	discoveryClient := discovery.NewDiscoveryClientForConfigOrDie(r)
+	clusterClient := typedcluster.NewForConfigOrDie(r)
+
+	clusterController, err := cluster.NewController(
+		crdClient,
+		discoveryClient,
+		clusterClient,
+		kcpSharedInformerFactory.Cluster().V1alpha1().Clusters(),
+		apiResourceClient,
+		kcpSharedInformerFactory.Apiresource().V1alpha1().APIResourceImports(),
+		*syncerImage,
+		kubeconfig,
+		resourcesToSync,
+		syncerMode,
+	)
 	if err != nil {
 		klog.Fatal(err)
 	}
-	clusterController.Start(numThreads)
 
 	apiresourceController, err := apiresource.NewController(
-		r,
+		apiResourceClient,
+		crdClient,
 		*autoPublishAPIs,
+		kcpSharedInformerFactory.Apiresource().V1alpha1().NegotiatedAPIResources(),
+		kcpSharedInformerFactory.Apiresource().V1alpha1().APIResourceImports(),
+		crdSharedInformerFactory.Apiextensions().V1().CustomResourceDefinitions(),
 	)
 	if err != nil {
 		klog.Fatal(err)
 	}
-	apiresourceController.Start(2)
 
-	go func() {
-		<-sigs
-		clusterController.Stop()
-		apiresourceController.Stop()
-	}()
+	kcpSharedInformerFactory.Start(ctx.Done())
+	kcpSharedInformerFactory.WaitForCacheSync(ctx.Done())
+
+	crdSharedInformerFactory.Start(ctx.Done())
+	crdSharedInformerFactory.WaitForCacheSync(ctx.Done())
+
+	clusterController.Start(numThreads, ctx.Done())
+	apiresourceController.Start(2, ctx.Done())
```

Comment:
If these took a ctx as well we could pipe that down through to [`process`](https://github.com/kcp-dev/kcp/blob/8522a6b6b42151540f54dfb16c3014445d7f798c/pkg/reconciler/cluster/controller.go#L265) and into [`reconcile`](https://github.com/kcp-dev/kcp/blob/8522a6b6b42151540f54dfb16c3014445d7f798c/pkg/reconciler/cluster/cluster.go#L26), instead of having [`context.TODO()`s](https://github.com/kcp-dev/kcp/blob/8522a6b6b42151540f54dfb16c3014445d7f798c/pkg/reconciler/cluster/controller.go#L278). WDYT?

Reply:
Yes yes yes! Thanks, I'll make that change too.

Reply:
Implemented everywhere I could do it easily. There's still a couple context.TODOs left, but can do in a follow-up.